File: src/test/java/com/regnosys/rosetta/common/serialisation/JsonLookupDataLoaderTest.java
package com.regnosys.rosetta.common.serialisation;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.regnosys.rosetta.common.serialisation.lookup.JsonLookupDataLoader;
import com.regnosys.rosetta.common.serialisation.lookup.LookupDataSet;
import org.junit.jupiter.api.Test;

import java.net.MalformedURLException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class JsonLookupDataLoaderTest {

    private static final Path RESOURCES_PATH = Paths.get("src/test/resources");

    private final ObjectMapper rosettaObjectMapper = RosettaObjectMapper.getNewRosettaObjectMapper();

    @Test
    void lookupsLoaded() throws MalformedURLException {
        // descriptor and input on same path
        Path path = RESOURCES_PATH.resolve("regs/test-reg-lookups");

        List<LookupDataSet> lookupDataSets = loadLookupDataSets(path, path);

        assertEquals(lookupDataSets.size(), 2);
        assertEquals(lookupDataSets.get(0).getName(), "IsExecutingEntityInvestmentFirm");
        assertEquals(lookupDataSets.get(0).getData().size(), 1);
        assertEquals(lookupDataSets.get(1).getName(), "ExecutingEntity");
        assertEquals(lookupDataSets.get(1).getData().size(), 4);
    }

    @Test
    void descriptorPathDoesNotExist() throws MalformedURLException {
        // descriptor and input on same path
        Path path = RESOURCES_PATH.resolve("not-found");

        List<LookupDataSet> lookupDataSets = loadLookupDataSets(path, path);

        assertEquals(lookupDataSets.size(), 0);
    }

    @Test
    void descriptorPathDoesNotDoesNotContainDescriptorFile() throws MalformedURLException {
        // descriptor and input on same path
        Path path = RESOURCES_PATH.resolve("test-workspaces");

        List<LookupDataSet> lookupDataSets = loadLookupDataSets(path, path);

        assertEquals(lookupDataSets.size(), 0);
    }

    private List<LookupDataSet> loadLookupDataSets(Path lookupDescriptorPath, Path inputPath) throws MalformedURLException {
        return new JsonLookupDataLoader(this.getClass().getClassLoader(),
                rosettaObjectMapper,
                lookupDescriptorPath.toUri().toURL(),
                Collections.singletonList(JsonLookupDataLoader.DEFAULT_DESCRIPTOR_NAME),
                inputPath.toUri().toURL()).load();
    }

    @SuppressWarnings("unused") // loaded at runtime
    static class ExecutingEntity {
        public Address addressOfBranch;
        public Address addressOfIncorporation;
        public Boolean isInvestmentFirm;
    }

    static class Address {
        public Country country;
    }

    static class Country {
        public String value;
    }
}

File: src/test/java/com/regnosys/rosetta/common/serialisation/JsonLookupDataLoaderTest.kt
package com.regnosys.rosetta.common.serialisation

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertSame

data class ExecutingEntity(val addressOfBranch: Address, val addressOfIncorporation: Address, val isInvestmentFirm: Boolean)
data class Address(val country: Country)
data class Country(val value: String)

internal class JsonLookupDataLoaderTest {
    private val rosettaObjectMapper = RosettaObjectMapper.getNewRosettaObjectMapper()

    @Test
    internal fun lookupsLoaded() {
        val lookupDataSets = JsonLookupDataLoader(this.javaClass.classLoader, rosettaObjectMapper, "regs/test-reg-lookups").load()
        assertEquals(lookupDataSets.size, 2)
        assertEquals(lookupDataSets[0].name, "IsExecutingEntityInvestmentFirm")
        assertEquals(lookupDataSets[0].data.size, 1)
        assertEquals(lookupDataSets[1].name, "ExecutingEntity")
        assertEquals(lookupDataSets[1].data.size, 4)
    }

    @Test
    internal fun descriptorPathDoesNotExist() {
        val lookupDataSets = JsonLookupDataLoader(this.javaClass.classLoader, rosettaObjectMapper, "not-found").load()
        assertSame(lookupDataSets.size, 0)
    }

    @Test
    internal fun descriptorPathDoesNotDoesNotContainDescriptorFile() {
        val lookupDataSets = JsonLookupDataLoader(this.javaClass.classLoader, rosettaObjectMapper, "test-workspaces").load()
        assertSame(lookupDataSets.size, 0)
    }
}